src/main/java/optic_fusion1/mcantimalware/Main.java

```diff
@@ -41,9 +42,18 @@
     public static void main(String[] args) {
         Main main = new Main();
         main.args = args;
-        main.mainThread = new Thread(main);
-        main.mainThread.setName("AntiMalware[Main Thread]");
-        main.mainThread.start();
+        ExecutorService executor = Executors.newCachedThreadPool(new ThreadFactory() {
+
+            private final AtomicInteger threadCount = new AtomicInteger(1);
+
+            @Override
+            public Thread newThread(Runnable r) {
+                Thread thread = new Thread(r);
+                thread.setName("MCAntiMalware/Check thread #" + threadCount.getAndIncrement());
+                return thread;
+            }
+        });
+        executor.execute(main);
```

Comment:
Wouldn't this cause the init process be be ran repeatedly?

Reply:
No. Its like the world loading in spigot 1.14.4 - u see the threads when its logging, right?

Reply:
Yes, HOWEVER i assume it will call, would it not?

`@Override
    public void run() {
        init();
    }
`

Reply:
Hmm.. lets do some tests

Reply:
init being the following

`  private void init() {
        AntiMalware.setMain(this);
        setupLogger();
        handleCommandLineArguments();
        downloadCheckDatabase(false);
        downloadChecksumDatabase(false);
        checkManager = new CheckManager();
        checkRegistery = new CheckRegistery(this);
        checkRegistery.registerChecks();
        realTimeScanner = new RealTimeScanner(this);
        if (scanSingleFile) {
            try {
                realTimeScanner.getScanner().scanFile(new File((String) options.valueOf("scanfile")));
            } catch (IOException ex) {
                if (shouldLogDebugMessages) {
                    logger.exception(ex);
                }
            }
            System.exit(0);
        }
        setupAutoUpdater();
        realTimeScanner.setupDirectoryWatcher();
    }
`

Reply:
Okay so this is what happened - i compiled it and i got it running on three plugins - LuckPerms, Vault and my own
https://gyazo.com/f0bee53c8057210c3226bc7ac62f45a5
It didn't create a bunch of threads, however, it freezed on vault.

Reply:
These are the 42 kb logs after i closed it
https://hasteb.in/nuhaleji.md

Reply:
Didn't freeze it's just scanning 24/7, three plugins also isn't enough to properly test a fix like this, you'd need a LARGE sample size, last time i tested this issue i ran it against 30k+ plugins

Reply:
future reference --debug true is a thing, it will give a lot more messages :P

Reply:
It looks like you're having another threadpool somewhere in the code i didn't realized, i will just go around and try finding it.

Reply:
Wouldn't be surprised if scanning needs rewritten _AGAIN_ just to support multithreading,

Reply:
Okay, now the system is multithreaded.

Reply:
Now it looks pretty good https://hasteb.in/isibuxoq.cs

Reply:
Gonna take a while to see if this fixes #191 need to redownload 30k plugins

Reply:
If it doesn't i have another idea which is creating for each check a thread. 98 threads this would cost some memory, however it won't freeze

Reply:
more than 98 threads actually, probably 100+

Reply:
39k+ resources downloaded so far

Reply:
How u download them wtf

Reply:
I made a program, allows me to do a variety of things, like auto-update the check database and in this case, mass download THOUSANDS of resources, even deletes skripts and corrupted jars

Reply:
Wow innnnnnnnnnnncredible

Reply:
40k resources atm, i could stop it here, but the skripts & corrupted jars will make scanning a pain in the ass.

and yea, i know.